src/katello/agent/katelloplugin.py

```diff
@@ -209,6 +210,70 @@ def validate_registration():
         raise
 
 
+class AgentRestart(object):
+    """
+    Restart the daemon after RPM upgrade.
+    The %post in the RPM will write the RESTART_FILE.  The recurring
+    'apply' action notices file and restarts the goferd service only when
+    this plugin is not longer busy.
+    """
+
+    COMMAND = 'service goferd restart'
+    RESTART_FILE = '/tmp/katello-agent-restart'
+
+    @staticmethod
+    def is_busy():
+        """
+        Determine if this plugin is busy by counting the pending
+        requests in the persistent queue.
+
+        :return: True if busy.
+        :rtype: bool
+        """
+        pending = plugin.scheduler.pending
+        path = os.path.join(pending.PENDING, pending.stream)
+        count = len(os.listdir(path))
+        return count > 0
+
+    @staticmethod
+    def touch():
+        """
+        Touch the RESTART_FILE.
+        """
+        fp = open(AgentRestart.RESTART_FILE, 'w+')
+        fp.close()
+
+    @staticmethod
+    def restart():
+        """
+        Restart the goferd service.
+          1. Delete the RESTART_FILE.
+          2. Restart
+          (restart failed)
+          3. Touch (create) the RESTART_FILE
+        """
+        os.unlink(AgentRestart.RESTART_FILE)
+        try:
+            p = Popen(AgentRestart.COMMAND, shell=True)
+            p.wait()
+        finally:
+            AgentRestart.touch()
```

Comment:
I think that the fire-and-forget way would be the best. I think anything that might fix the issue would happen after one retry, and continuing indefinitely sounds like it'd be less desirable, as wouldn't stop the cycle at any point.